### PR TITLE
Fix/risk slider interaction

### PR DESCRIPTION
**Pull request by rhubarbgarden**
- Predicted risk throws if negative
- Risk is zero only when an account has no exposure (liabilities) at all

### CODE REVIEW

File: libraries/ts/src/margin/marginAccount.ts [hunk 1]
```diff
@@ -120,11 +120,38 @@ export class MarginAccount {
...
+
+  /** A just-okay risk indicator (TODO improve me) */
+  computeRiskIndicator(
+    requiredCollateral: number,
+    weightedCollateral: number,
+    liabilities: number,
+  ): number {
+    if (requiredCollateral < 0) throw "requiredCollateral must be non-negative"
+    if (weightedCollateral < 0) throw "weightedCollateral must be non-negative"
+    if (liabilities < 0) throw "liabilities must be non-negative"
```
Comment (Tristyn) [inline libraries/ts/src/margin/marginAccount.ts:146]: It's good hygiene to throw something deriving Error
```suggestion
    if (requiredCollateral < 0) throw new Error("requiredCollateral must be non-negative")
    if (weightedCollateral < 0) throw new Error("weightedCollateral must be non-negative")
    if (liabilities < 0) throw new Error("liabilities must be non-negative")
```

File: libraries/ts/src/margin/marginAccount.ts [hunk 2]
```diff
@@ -572,10 +599,10 @@ export class MarginAccount {
       }
     }
 
-    const effectiveCollateral = weightedCollateral.sub(exposure)
+    const effectiveCollateral = weightedCollateral.sub(liabilities)
 
     return {
-      exposure,
+      liabilities: liabilities,
```
Comment (Tristyn) [inline libraries/ts/src/margin/marginAccount.ts:605]:
```suggestion
      liabilities,
```

File: libraries/ts/src/margin/pool/pool.ts [hunk 3]
```diff
@@ -982,13 +982,11 @@ export class Pool {
   }
 
   /// Projects the deposit and borrow rates after a deposit into the pool.
-  projectAfterDeposit(marginAccount: MarginAccount, amount: TokenAmount): PoolProjection {
-    if (this.info == undefined) {
-      throw "must have pool info initialised"
-    }
+  projectAfterDeposit(marginAccount: MarginAccount, amount: number): PoolProjection {
+    if (this.info === undefined) throw "must have info"
```
Comment (Tristyn) [inline libraries/ts/src/margin/pool/pool.ts:986]:
```suggestion
    if (this.info === undefined) throw new Error("must have info")
```

File: libraries/ts/src/margin/pool/pool.ts [hunk 4]
```diff
@@ -997,30 +995,36 @@ export class Pool {
...
     return { riskIndicator, depositRate, borrowRate }
   }
 
   /// Projects the deposit and borrow rates after a withdrawal from the pool.
-  projectAfterWithdraw(marginAccount: MarginAccount, amount: TokenAmount): PoolProjection {
-    if (this.info == undefined) {
-      throw "must have pool info initialised"
-    }
+  projectAfterWithdraw(marginAccount: MarginAccount, amount: number): PoolProjection {
+    if (this.info === undefined) throw "must have info"
```
Comment (Tristyn) [inline libraries/ts/src/margin/pool/pool.ts:1015]:
```suggestion
    if (this.info === undefined) throw new Error("must have info")
```

File: libraries/ts/src/margin/pool/pool.ts [hunk 5]
```diff
@@ -997,30 +995,36 @@ export class Pool {
...
 
-    if (amount.tokens > this.vault.tokens) {
-      throw "not enough tokens in the vault"
-    }
+    const symbol = this.symbol
+    if (symbol === undefined) throw "must have symbol"
```
Comment (Tristyn) [inline libraries/ts/src/margin/pool/pool.ts:1018]:
```suggestion
    if (symbol === undefined) throw new Error("must have symbol")
```

File: libraries/ts/src/margin/pool/pool.ts [hunk 6]
```diff
@@ -997,30 +995,36 @@ export class Pool {
...
+
+    const position = marginAccount.poolPositions[symbol]
+    if (position === undefined) throw "must have position"
```
Comment (Tristyn) [inline libraries/ts/src/margin/pool/pool.ts:1021]:
```suggestion
    if (position === undefined) throw new Error("must have position")
```

File: libraries/ts/src/margin/pool/pool.ts [hunk 7]
```diff
@@ -997,30 +995,36 @@ export class Pool {
...
+    
+    // G1, referenced below
+    if (amount > position.depositBalance.tokens) throw "amount can't exceed deposit"
```
Comment (Tristyn) [inline libraries/ts/src/margin/pool/pool.ts:1024]:
```suggestion
    if (amount > position.depositBalance.tokens) throw new Error("amount can't exceed deposit")
```

File: libraries/ts/src/margin/pool/pool.ts [hunk 8]
```diff
@@ -1029,26 +1033,29 @@ export class Pool {
...
     return { riskIndicator, depositRate, borrowRate }
   }
 
   /// Projects the deposit and borrow rates after a borrow from the pool.
-  projectAfterBorrow(marginAccount: MarginAccount, amount: TokenAmount): PoolProjection {
-    if (this.info == undefined) {
-      throw "must have pool info initialised"
-    }
+  projectAfterBorrow(marginAccount: MarginAccount, amount: number): PoolProjection {
+    if (this.info === undefined) throw "must have info"
```
Comment (Tristyn) [inline libraries/ts/src/margin/pool/pool.ts:1055]:
```suggestion
    if (this.info === undefined) throw new Error("must have info")
```

File: libraries/ts/src/margin/pool/pool.ts [hunk 9]
```diff
@@ -1057,31 +1064,38 @@ export class Pool {
...
     return { riskIndicator, depositRate, borrowRate }
   }
 
   /// Projects the deposit and borrow rates after repaying a loan from the pool.
-  projectAfterRepay(marginAccount: MarginAccount, amount: TokenAmount): PoolProjection {
-    if (this.info == undefined) {
-      throw "must have pool info initialised"
-    }
+  projectAfterRepay(marginAccount: MarginAccount, amount: number): PoolProjection {
+    if (this.info === undefined) throw "must have info"
```
Comment (Tristyn) [inline libraries/ts/src/margin/pool/pool.ts:1086]:
```suggestion
    if (this.info === undefined) throw new Error("must have info")
```

File: libraries/ts/src/margin/pool/pool.ts [hunk 10]
```diff
@@ -1057,31 +1064,38 @@ export class Pool {
...
 
-    if (amount.tokens > this.borrowedTokens.tokens) {
-      throw "not enough borrowed tokens"
-    }
+    const symbol = this.symbol
+    if (symbol === undefined) throw "must have symbol"
```
Comment (Tristyn) [inline libraries/ts/src/margin/pool/pool.ts:1089]:
```suggestion
    if (symbol === undefined) throw new Error("must have symbol")
```

**Review by Tristyn — COMMENTED** [review]
Can you go through the codebase and throw `new Error(string)` instead of `string`. It must derive from Error to have a stacktrace

Searching for `throw "` would find them all

File: libraries/ts/src/margin/pool/pool.ts [hunk 11]
```diff
@@ -982,13 +982,11 @@ export class Pool {
...
-      throw "must have pool info initialised"
-    }
+  projectAfterDeposit(marginAccount: MarginAccount, amount: number): PoolProjection {
+    if (this.info === undefined) throw "must have info"
 
     const borrowedTokens = this.borrowedTokens.tokens
-    const totalValue = this.totalValue.add(amount).tokens
+    const totalValue = this.totalValue.tokens + amount
 
     const utilRatio = borrowedTokens / totalValue
```
Comment (Tristyn) [inline libraries/ts/src/margin/pool/pool.ts:991]: More accurate name?

```suggestion
    const totalTokens = this.totalValue.tokens + amount

    const utilRatio = borrowedTokens / totalTokens
```

Comment (eyevz) [reply libraries/ts/src/margin/pool/pool.ts]: Yup agreed

**Review by davodesign84 — DISMISSED** [review]

**Review by davodesign84 — APPROVED** [review]

**Review by Tristyn — APPROVED** [review]

File: libraries/ts/src/margin/marginAccount.ts [hunk 12]
```diff
@@ -68,7 +68,7 @@ export interface AccountSummary {
 }
 
 export interface Valuation {
-  exposure: Number128
```
[the remark this answers is not in the document]
Comment (davodesign84) [reply libraries/ts/src/margin/marginAccount.ts:71]: would it not make sense to do it as part of this PR?

**Pull request merged**